enhancements/operator-conditions.md

```diff
@@ -0,0 +1,188 @@
+  - "N/A"
+---
+
+# operator-conditions
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to allow operators managed by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) operator to communicate conditions that OLM would otherwise be unable to infer. In particular, these conditions will be used by OLM to prevent irreversible actions (such as upgrades) when the operator is in a state that would be problematic.
+
+## Motivation
+
+As part of its role in managing the lifecycle of an operator, OLM infers the state of an operator from the state of Kubernetes resources that define the operator. While this sanity-check provides some level of assurance that an operator is in a given state, there are many complex states specific to each operator that OLM is unable to infer on its own. OLM could better handle the lifecycle of operators by providing them the means to communicate complex states. In particular, there are a number of benefits if an operator managed by OLM could communicate when it should not be upgraded. Consider the following scenario:
+
+>An operator installed with automatic updates is currently performing a migration that could result in data loss if OLM upgrades the operator. While this migration process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the migration process to finish, resulting in unrecoverable data loss. Had OLM given the operator the ability to communicate that it should not be upgraded, no data loss would have occurred.
+
+As shown in the example above, by allowing operators to report complex condition and influence OLM's management, the overall user experience could be improved for both operator authors and OLM users.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate conditions that influence OLM behavior.
+- Allow the list of "OLM Supported Conditions" to grow in the future.
+- Update OLM to support the "Upgrade Readiness" condition defined above.
+- Allow operators to communicate additional information to admins without influencing OLM behavior.
+- Preserve existing behavior if the operator does not take advantage of this feature.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want operators managed by OLM to be able to communicate complex states that OLM cannot infer on its own.
+
+#### Story 2
+
+As an operator author, I want operators managed by OLM to be able to communicate whether or not they are ready to be upgraded.
+
+## Design Details
+
+### OperatorHealth CRD
```

Comment:
I know that @ecordell has raised concerned over the CRD name. Let's brainstorm here.

Reply:
OperatorReport

Reply:
OperatorCondition

Reply:
Could this just be `Condition` since we will already have the `operator` part in the group name?

Reply:
OperatorMessages

Reply:
OperatorState

Reply:
OperatorAnnouncement

Reply:
how about OperatorMood, OperatorTemper or OperatorConstitution ?

Reply:
`Condition` seems like the right thing

Reply:
off-topic: `Condition` [singular] makes me feel like the operator should create/update/delete a CR per distinct condition type, then `Conditions` [plural] is just a list operation.

After stating this I also considered that we originally wanted a [per operator] singleton as a control-point for probing behavior (push vs. pull) discussed on the supplanted enhancement -- the above proposal scatters that control point.

Reply:
I think `Condition` is fine if we position it as "the condition of the operator".

Reply:
I am not opposed to `Conditions` though.

Reply:
condition is a bit confusing IMO - it may be interpreted as a state that needs to occur instead of a state that is.(install condition vs conditional install)

Reply:
@gallettilance based on the accepted use of [conditions](https://github.com/kubernetes/apimachinery/blob/9254095ca5cab3666d500ec67cd00f9ab0d113d7/pkg/apis/meta/v1/types.go#L1367) in kubernetes I think this shouldn't be a huge concern.

Reply:
Updated the PR to rename the CRD to Condition.